### PR TITLE
Enhancements to CA1801 fixer

### LINKED ISSUE #4697: CA1801 - Remove unused parameter code fix should not attempt to fix custom operators
For the following fragment:

```csharp
public struct Unit
{
    public static bool operator ==(Unit x, Unit y) => true;
    public static bool operator !=(Unit x, Unit y) => false;
}
```

CA1801 suggests to remove the parameters, resulting in invalid code with

```
CS1534 - Overloaded binary operator takes two parameters.
```

The same holds for unary operators of course. While the warning for CA1801 is good, it's unactionable in terms of a code fix for special members like operators.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4697

Commit-by-commit review should be easier.

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs [hunk 1]
```diff
@@ -511,6 +511,18 @@ class MyValue
...
+        public async Task Operators_CSharp()
+        {
+            var code = @"
+public struct Unit
+{
+    public static bool operator ==(Unit [|x|], Unit [|y|]) => true;
+    public static bool operator !=(Unit [|x|], Unit [|y|]) => false;
+}";
+            await VerifyCS.VerifyCodeFixAsync(code, code);
+        }
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs:524]: This test is failing with:

```
    Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Iterative code fix application
    Expected '0' iterations but found '1' iterations.
```

Not sure what this means or how to fix that.

AFAIK, `VerifyCS.VerifyCodeFixAsync(code, code)` verifies the diagnostics and verifies that no code fix is offered.

Comment (Youssef1313) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs:524]: Ok after reading the fixer code, it sounds like this is *an existing bug*, where a codefix is offered but it does "nothing":

image

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs [hunk 2]
```diff
@@ -511,6 +511,25 @@ class MyValue
...
+    public static bool operator ==(Unit [|x|], Unit [|y|]) => true;
+    public static bool operator !=(Unit [|x|], Unit [|y|]) => false;
+}";
+
+            var fixedCode = @"
+public struct Unit
+{
+    public static bool operator ==(Unit _, Unit _1) => true;
+    public static bool operator !=(Unit _, Unit _1) => false;
+}";
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs:529]: The fix all is failing for this, it produces `_` for all parameters.
Can someone give me a guidance on how to fix this?

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs [hunk 3]
```diff
@@ -28,19 +31,61 @@ public sealed override FixAllProvider GetFixAllProvider()
...
             {
+                SyntaxNode diagnosticNode = root.FindNode(diagnostic.Location.SourceSpan);
+                SyntaxNode declarationNode = GetParameterDeclarationNode(diagnosticNode);
+                if (semanticModel.GetDeclaredSymbol(declarationNode, cancellationToken) is not IParameterSymbol parameterSymbol)
+                {
+                    Debug.Fail("Symbol was not IParameterSymbol.");
+                    continue;
+                }
+
+                if (parameterSymbol.ContainingSymbol is not IMethodSymbol methodSymbol)
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs:50]: I needed an `IMethodSymbol` to loop on its parameter so I could properly rename them as `_`, `_1`, etc. Before merging, I'll need to add a test for positional records to see whether it fails, and fix it appropriately. Or totally find another better way.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs [hunk 4]
```diff
@@ -28,19 +31,61 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+
+            // Use _, _1, _2, etc.
+            var newName = "_";
+            var suffix = 0;
+            while (parameterNames.Contains(newName))
+            {
+                newName = $"_{++suffix}";
+            }
+
+            return await Renamer.RenameSymbolAsync(document.Project.Solution, parameterSymbol, newName, document.Project.Solution.Options, cancellationToken).ConfigureAwait(false);
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs:88]: This doesn't handle a case like the following:

```csharp
private string M(string s) // s should be renamed to _1 here. This is not handled currently.
{
    var _ = Console.ReadLine(); // note this is an actual variable name, not a discard.
    return _;
}
```

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs [hunk 5]
```diff
@@ -28,19 +31,61 @@ public sealed override FixAllProvider GetFixAllProvider()
...
+                    // See https://github.com/dotnet/roslyn-analyzers/issues/1466
+                    context.RegisterCodeFix(
+                        new MyCodeAction(
+                            MicrosoftCodeQualityAnalyzersResources.RemoveUnusedParameterMessage,
+                            async ct => await RemoveNodes(document, declarationNode, parameterSymbol, ct).ConfigureAwait(false),
+                            equivalenceKey: MicrosoftCodeQualityAnalyzersResources.RemoveUnusedParameterMessage),
+                        diagnostic);
+                }
+
                 context.RegisterCodeFix(
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.Fixer.cs:67]: When registering multiple codefixes, is it guaranteed they'll appear in the same order in the bulb-light window?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs [hunk 6]
```diff
@@ -385,7 +385,7 @@ public void Caller()
             var fix = @"
 class C
 {
-    public static explicit operator int(C [|value|]) => 0;
+    public static explicit operator int(C _) => 0;
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParametersTests.Fixer.cs:388]: Does the test library support selecting different code actions in the same test?

**Comment by mavasani** [comment]
Ping @Youssef1313 about the last question

**Comment by Youssef1313** [comment]
Sorry @mavasani, I thought I had already commented on this.

Given this is a codefix-only issue (and not very serious), it shouldn't be urgent IMO to fix in .NET 5. I'm going close.

**Pull request closed without merging**